op-service/sources/eth_client.go

```diff
@@ -354,24 +314,21 @@ func (s *EthClient) PayloadByLabel(ctx context.Context, label eth.BlockLabel) (*
 func (s *EthClient) FetchReceipts(ctx context.Context, blockHash common.Hash) (eth.BlockInfo, types.Receipts, error) {
 	info, txs, err := s.InfoAndTxsByHash(ctx, blockHash)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, fmt.Errorf("querying block: %w", err)
 	}
-	// Try to reuse the receipts fetcher because is caches the results of intermediate calls. This means
-	// that if just one of many calls fail, we only retry the failed call rather than all of the calls.
-	// The underlying fetcher uses the receipts hash to verify receipt integrity.
```

Comment:
How did we preserve this behaviour where if we get some receipts successfully then get an error, the next time the derivation pipeline requests those same receipts it doesn't have to re-request the receipts it already has?  A block based cache won't provide this because you can only cache the receipts once you have them all for a block and that may require multiple calls (eg for the basic rpc kind).

Reply:
If it is only basic that this applies to, maybe it should just cache individual receipts by tx hash as a simple solution.

Reply:
This behavior was indeed not preserved. I added it back in 52673275d7ea13289c1978b4aa42d8133a909afa. I also added better synchronization of fetching jobs to avoid the creation of duplicate fetching jobs. In the old implementation, if there were two racing concurrent accesses to the fetching jobs cache, there might have been created two concurrent fetching jobs for the same block hash.